pkg/jobmanager/job.go

```diff
@@ -122,6 +90,73 @@ func newJob(ctx xcontext.Context, registry *pluginregistry.PluginRegistry, jobDe
 	return &job, nil
 }
 
+func buildTestsFromDescriptors(
+	ctx xcontext.Context,
+	registry *pluginregistry.PluginRegistry,
+	testDescriptors []*test.TestDescriptor,
+	stepsDescriptors []test.TestStepsDescriptors) ([]*test.Test, error) {
+
+	tests := make([]*test.Test, 0, len(testDescriptors))
+
+	for index, td := range testDescriptors {
+		thisTestStepsDescriptors := stepsDescriptors[index]
+
+		if err := td.Validate(); err != nil {
+			return nil, fmt.Errorf("could not validate test descriptor: %w", err)
+		}
+		bundleTargetManager, err := registry.NewTargetManagerBundle(td)
+		if err != nil {
+			return nil, err
+		}
+		bundleTestFetcher, err := registry.NewTestFetcherBundle(ctx, td)
+		if err != nil {
+			return nil, err
+		}
+
+		bundleTest, err := newBundlesFromSteps(ctx, thisTestStepsDescriptors.TestSteps, registry)
+		if err != nil {
+			return nil, fmt.Errorf("could not create test steps bundles: %w", err)
+		}
+		testName := thisTestStepsDescriptors.TestName
+		if err := limits.NewValidator().ValidateTestName(testName); err != nil {
+			return nil, err
+		}
+
+		var (
+			bundleCleanup []test.TestStepBundle
+			cleanupName   string
+		)
+		if len(thisTestStepsDescriptors.CleanupSteps) > 0 {
+			bundleCleanup, err = newBundlesFromSteps(ctx, thisTestStepsDescriptors.CleanupSteps, registry)
+			if err != nil {
+				return nil, fmt.Errorf("could not create cleanup test steps bundles: %w", err)
+			}
+			cleanupName = thisTestStepsDescriptors.CleanupName
+			if err := limits.NewValidator().ValidateTestName(cleanupName); err != nil {
+				return nil, err
+			}
+		}
+		if err := validateNoDuplicateLables(bundleTest, bundleCleanup); err != nil {
```

Comment:
you can remove duplication here (and an extra loop in func) by saying
```suggestion
		if err := validateNoDuplicateLables(append(bundleTest, bundleCleanup...)); err != nil {
```